### ISSUE TITLE
GRDB 5.26.0 cost more time on value Observation

### ISSUE THREAD

**Issue by karthuszY** [groue/GRDB.swift#1256]
### What did you do?

I update GRDB from 5.24.0 to 5.26.0. And then I feel my App got not smooth.

### Environment

**GRDB flavor(s):**   GRDB.swift/SQLCiphe
**GRDB version:** 5.26.0
**Installation method:** `pod 'GRDB.swift/SQLCipher', :git => 'git@github.com:groue/GRDB.swift.git'`
**Xcode version:** Version 13.0 (13A233)
**Swift version:** Swift 5
**Platform(s) running GRDB:** iOS
**macOS version running Xcode:** 12.0.1 (21A559)

### Related
GRDB **5.24.0** open cost：

![image](https://user-images.githubusercontent.com/13864988/182755224-59d2648e-45f5-43ba-8aa5-64d7e6652966.png)

GRDB **5.26.0** open cost：

![image](https://user-images.githubusercontent.com/13864988/182755349-671e4dc5-ce5f-4baf-9053-45f5bc583fb8.png)

**Comment by groue** [comment]
Hello @karthuszY,

Do you happen to use a DatabaseQueue, or DatabasePool?

Also, I won't be able to look at the issue until mid of August. Meanwhile, it would be useful if you would extend the issue with enough detail so that it is possible to **reproduce** the problem.

**Comment by karthuszY** [comment]
I use DatabasePool.

Here is the code:

```
       appInfoCancelable?.cancel()
        chainInfoCancelable?.cancel()
        
        let appInfoObserveation = ValueObservation.tracking(AppInfoModel
                                                                .filter(AppInfoModel.Columns.id == appId)
                                                                .fetchOne)
        appInfoCancelable = appInfoObserveation.start(in: db,
                                                      scheduling: .immediate) { error in
            
        } onChange: { [weak self] appInfo in
       
        }
        
        let chainInfoObserveation = ValueObservation.tracking(ChainInfoModel.fetchAll(_:))
        chainInfoCancelable = chainInfoObserveation.start(in: db,
                                                          scheduling: .immediate,
                                                          onError: { error in
            
        }, onChange: { [weak self] _ in
        })
```

**Comment by groue** [comment]
Hello @karthuszY,

I can't reproduce any significant performance difference. Is it possible for you to setup a minimal sample project?

**Comment by groue** [comment]
No news is good news! Feel free to reopen when it is possible to provide some more detailed feedback :)

**Comment by rpassis** [comment]
FWIW this is actually a problem we are also experiencing when upgrading from 5.25.0 to 5.26.1 (also using `DatabasePool`).
I changed it `DatabaseQueue` and the problem goes away. We narrowed it down to [this PR](https://github.com/groue/GRDB.swift/pull/1248). 

I'll try and get a minimum reproducible and also see if we can figure out exactly what's going on.

**Comment by groue** [comment]
This will be greatly appreciated, @rpassis 👍

**Comment by rpassis** [comment]
Hi @groue just a quick update that I don't have a reproducible example yet but have been able to capture a bit more information that might be helpful in debugging this. 

Firstly, I can only reproduce this when using SQLCipher alongside a DatabasePool configured with a passphrase i.e. 

```swift
config.prepareDatabase { db in
    try db.usePassphrase(passphraseData)
}
```

It looks like while `ValueObservation` tracking is running, a computationally expensive operation is blocking the main thread. In our case this usually lasts several seconds and the UI remains non-responsive. I was able to capture the stack trace below while the main thread was blocked:

![CleanShot 2023-03-15 at 15 55 50@2x](https://user-images.githubusercontent.com/387079/225219870-3ed66214-a48e-4420-ad68-ce2f28882370.png)

I'll keep digging and will update this thread with any new findings. 

Thank you!
Rog

**Comment by groue** [comment]
> It looks like while `ValueObservation` tracking is running, a computationally expensive operation is blocking the main thread.

`ValueObservation` only performs database operations on the main thread if the application instructs it to do so by starting the observation with the `.immediate` scheduling (see [ValueObservation Scheduling](https://swiftpackageindex.com/groue/grdb.swift/documentation/grdb/valueobservation#ValueObservation-Scheduling)).

When database operations are slow, this blocks the main thread for a noticeable delay. I hope it's clear that there is no magic here: to notify its initial value, `ValueObservation` can't be faster than the underlying SQLite requests.

To restore UI responsiveness, an application may make the database operations much faster. For example, could the request be enhanced with an [index](https://en.wikipedia.org/wiki/Database_index)? When the request is very fast, you can keep running it on the main thread: the user won't notice.

Another way to avoid blocking the user interface is to stop performing the slow database operations on the main thread. Don't start the `ValueObservation` with the `.immediate` scheduling (which asks for blocking the main thread until the initial value is fetched), and prefer the default scheduling. Values are still notified on the main thread (so you can display them), but SQLite operations are not performed on the main thread (so that the app remains responsive). The app can display a loading screen until the first value is notified.

---

The above paragraphs do not describe a defect or a regression. It is the normal way to deal with `ValueObservation`: simple and easy `.immediate` scheduling for fast requests, and default asynchronous scheduling for slow requests (with a loading screen or any other UI until the slow I/O are completed).

In other words, GRDB can't be faster than SQLite/SQLCipher.

If this GitHub issue describes an actual GRDB performance regression, then I'm very curious to learn about it. But it is important to be focused on actual GRDB operations, not SQLite ones. I understand it is not easy, and I'm very thankful for the time spent.

**Comment by rpassis** [comment]
I think what you said above all makes sense however what I can't make sense of is why this behaviour is not present under `5.25.0` but is pretty evident starting with `5.26.0`.

**Comment by groue** [comment]
What about profiling your app with Instruments.app, and looking for the GRDB api that starts performing worse? You could even share actual profiling information.

**Comment by groue** [comment]
With Instruments, you could see what is slow. In this screenshot, `ValueObservation.start` is so fast that it barely details where time is spent. But in your app, you may discover something interesting.

<img width="357" alt="Capture d’écran 2023-03-15 à 13 11 36" src="https://user-images.githubusercontent.com/54219/225306224-19f502ea-2b18-4e3d-9ccd-0fde817d71d7.png">

<img width="836" alt="Capture d’écran 2023-03-15 à 13 15 58" src="https://user-images.githubusercontent.com/54219/225306544-34e8268f-56a9-41c4-82eb-f83eadcded90.png">

<img width="1249" alt="Capture d’écran 2023-03-15 à 13 08 15" src="https://user-images.githubusercontent.com/54219/225304703-dd143f46-14c2-48e0-a74b-c93a1adfeca6.png">

**Comment by rpassis** [comment]
Hi @groue here's a sample of the time profile intrumentation showing the same stack trace as you. You can see the call to `ValueObservation.start` in this sample takes ~450ms when running with GRDB `5.26.1`. Also note the CPU usage peaking at 100%:

![CleanShot 2023-03-16 at 09 21 05@2x](https://user-images.githubusercontent.com/387079/225465542-a0986e93-c9a1-476f-b634-7e2a62a4b1e7.png)

When I downgrade it back to `5.25.0` then I start seeing a sample that is more similar to the one you posted above:

![CleanShot 2023-03-16 at 09 45 38@2x](https://user-images.githubusercontent.com/387079/225469159-f2400f0a-a8ea-49fb-b933-bc7fb842c74c.png)

I'd be happy to share (privately) the Instruments file that contains both runs if you think it might be helpful, just let me know how to do it.

Thanks again,
Rog

**Comment by groue** [comment]
> I'd be happy to share (privately) the Instruments file that contains both runs if you think it might be helpful, just let me know how to do it.

Your first screenshot shows that quite some time (448ms !!!) is spent while creating a `DatabaseSnapshot`, totally dominated by SQLCipher.

I'm confident this is the change we were looking for, because GRDB 5.25.0 would use a distinct technique. 🎉

You're paying a hard price for using SQLCipher 😅

OK, we can reopen the issue. I'll look for a workaround - probably by letting the app opt out of #1248. In exchange, you'll get a double notification of the initial value 😬. I don't see any other choice, because waiting for SQLCipher up to 0.5 second before you can display the initial value, even without blocking the main thread, is not acceptable.

**Comment by groue** [comment]
@rpassis, @karthuszY, I have all reasons to think that #1350 resolves the performance regression.

It would be very helpful from you to test this branch 🙏

**Comment by groue** [comment]
@rpassis, I just pushed some enhancement to #1350. OK now I stand still and wait for your feedback :-) Don't forget to [upgrade](https://github.com/groue/GRDB.swift/blob/master/Documentation/GRDB6MigrationGuide.md) to GRDB 6 if you are still running GRDB 5!

**Comment by rpassis** [comment]
Thanks so much @groue I'll give this a good test early next week and let you know.

Because we are using SQLCipher I'm not sure I'll be able to test that the firing of double values remains functional. Whenever I try to build GRDB/SQLCipher with `SQLITE_ENABLE_SNAPSHOT` I get the compiler complaining about missing symbols:

```
❌ Undefined symbols for architecture arm64
❌   "_sqlite3_snapshot_cmp", referenced from:
❌   "_sqlite3_snapshot_free", referenced from:
❌   "_sqlite3_snapshot_get", referenced from:
❌   "_sqlite3_snapshot_open", referenced from:
❌ ld: symbol(s) not found for architecture arm64
❌ clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Is that because the version of SQL used by SQLCipher does not support snapshots?
Thanks again!
Rog

**Comment by groue** [comment]
Fixed by #1350 :-)
